src/circuit/boolean.rs

```diff
@@ -523,6 +523,278 @@ impl Boolean {
+                // equals
+                // (a and c)
+                return Boolean::and(
+                    cs,
+                    a,
+                    c
+                );
+            },
+            (a, b, &Boolean::Constant(false)) => {
+                // If c is false,
+                // (a and b) xor (a and c) xor (b and c)
+                // equals
+                // (a and b)
+                return Boolean::and(
+                    cs,
+                    a,
+                    b
+                );
+            },
+            (a, b, &Boolean::Constant(true)) => {
+                // If c is true,
+                // (a and b) xor (a and c) xor (b and c)
+                // equals
+                // (a and b) xor (a) xor (b)
+                // equals
+                // not ((not a) and (not b))
+                return Ok(Boolean::and(
+                    cs,
+                    &a.not(),
+                    &b.not()
+                )?.not());
+            },
+            (a, &Boolean::Constant(true), c) => {
+                // If b is true,
+                // (a and b) xor (a and c) xor (b and c)
+                // equals
+                // (a) xor (a and c) xor (c)
+                return Ok(Boolean::and(
+                    cs,
+                    &a.not(),
+                    &c.not()
+                )?.not());
+            },
+            (&Boolean::Constant(true), b, c) => {
+                // If a is true,
+                // (a and b) xor (a and c) xor (b and c)
+                // equals
+                // (b) xor (c) xor (b and c)
+                return Ok(Boolean::and(
+                    cs,
+                    &b.not(),
+                    &c.not()
+                )?.not());
+            },
+            (&Boolean::Is(_), &Boolean::Is(_), &Boolean::Is(_)) |
+            (&Boolean::Is(_), &Boolean::Is(_), &Boolean::Not(_)) |
+            (&Boolean::Is(_), &Boolean::Not(_), &Boolean::Is(_)) |
+            (&Boolean::Is(_), &Boolean::Not(_), &Boolean::Not(_)) |
+            (&Boolean::Not(_), &Boolean::Is(_), &Boolean::Is(_)) |
+            (&Boolean::Not(_), &Boolean::Is(_), &Boolean::Not(_)) |
+            (&Boolean::Not(_), &Boolean::Not(_), &Boolean::Is(_)) |
+            (&Boolean::Not(_), &Boolean::Not(_), &Boolean::Not(_))
+            => {}
+        }
+
+        let maj = cs.alloc(|| "maj", || {
+            maj_value.get().map(|v| {
+                if *v {
+                    E::Fr::one()
+                } else {
+                    E::Fr::zero()
+                }
+            })
+        })?;
+
+        // ¬(¬a ∧ ¬b) ∧ ¬(¬a ∧ ¬c) ∧ ¬(¬b ∧ ¬c)
+        // (1 - ((1 - a) * (1 - b))) * (1 - ((1 - a) * (1 - c))) * (1 - ((1 - b) * (1 - c)))
+        // (a + b - ab) * (a + c - ac) * (b + c - bc)
+        // -2abc + ab + ac + bc
+        // a (-2bc + b + c) + bc
+        //
+        // (b) * (c) = (bc)
+        // (2bc - b - c) * (a) = bc - maj
```

Comment:
There's another way of doing this which *might* be more efficient; it depends to what extent "= 0" constraints can be optimized. The majority function can be written as a + b + c >= 2. Let a + b + c = 2.r + s with r, s boolean. Then substitute s = a + b + c - 2.r into the boolean constraint of s, giving:

    (1 - r) × (r) = 0
    (1 - (a + b + c - 2.r)) × (a + b + c - 2.r) = 0

Reply:
(I like this because it brings out the symmetry, but obviously that isn't a good enough reason to use it unless it's more efficient.)

Reply:
Cool suggestion, but I don't think that in practice this would benefit performance. The fact the C term becomes zero doesn't help performance in Groth16. In fact, this would likely harm performance since all of the variables will appear in the B term, which will result in increased density for the B query (and increased G2 arithmetic).